File: core/sanitizer.py
"""이그레스 가드 — 조립된 VulnFact를 policy/egress-policy.json 에 대고 검증한다.

3중 검증이며, 하나라도 걸리면 **전송을 중단한다(fail-closed)**:

  1. 스키마 검증 — 모든 키가 허용 목록에 있는가, 금칙 필드명이 아닌가
  2. 값 검증   — 타입·범위·enum·정규식·길이를 통과하는가
  3. 금칙 패턴 — IP·경로·이메일·MAC·UUID·내부 도메인·한글·자격증명 흔적이 없는가

같은 정책 파일을 web/js/core/sanitizer.js 도 읽으며,
policy/egress-test-vectors.json 으로 두 구현의 동치성을 검증한다.

이 가드는 2차 방어다. 1차 방어는 core/vulnfact.py 의 함수 시그니처가
로컬 계층을 아예 받지 않는다는 사실이다.
"""

from __future__ import annotations

import json
import re
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any

from .config import Config, get_config
from .policy import LoadedPolicy, load as load_policy


def _flags(spec: str) -> int:
    """정책의 flags 문자열을 re 플래그로 옮긴다. JS의 RegExp 플래그와 같은 표기."""
    value = 0
    if "i" in spec:
        value |= re.IGNORECASE
    if "m" in spec:
        value |= re.MULTILINE
    if "s" in spec:
        value |= re.DOTALL
    return value


class EgressBlocked(Exception):
    """검증 실패로 전송이 중단되었다. 어떤 규칙에 왜 걸렸는지를 담는다."""

    def __init__(self, violations: list["Violation"]):
        self.violations = violations
        super().__init__(
            "외부 전송이 차단되었습니다 ("
            + "; ".join(f"{v.rule}@{v.path}: {v.reason}" for v in violations[:5])
            + (f" 외 {len(violations) - 5}건" if len(violations) > 5 else "")
            + ")"
        )


@dataclass(frozen=True)
class Violation:
    rule: str            # forbidden_field / unknown_field / type / pattern / range / forbidden_pattern
    path: str            # facts[0].cvss_vector 처럼 어디서 걸렸는지
    reason: str
    sample: str = ""     # 걸린 값의 일부. 로그로 남으므로 짧게 자른다.

    def to_dict(self) -> dict[str, str]:
        return {"rule": self.rule, "path": self.path, "reason": self.reason, "sample": self.sample}


@dataclass
class SanitizeResult:
    ok: bool
    facts: list[dict[str, Any]] = field(default_factory=list)
    violations: list[Violation] = field(default_factory=list)
    policy_version: str = ""
    policy_sha256: str = ""

    def to_dict(self) -> dict[str, Any]:
        return {
            "ok": self.ok,
            "facts": self.facts,
            "violations": [v.to_dict() for v in self.violations],
            "policy_version": self.policy_version,
            "policy_sha256": self.policy_sha256,
        }


class EgressGuard:
    def __init__(self, policy: LoadedPolicy):
        self.policy = policy
        data = policy.data
        self.allowed: dict[str, Any] = data.get("allowed_fields") or {}
        self.forbidden_names: set[str] = {n.lower() for n in data.get("forbidden_field_names") or ()}
        self.limits: dict[str, Any] = data.get("limits") or {}
        # 정규식 플래그는 정책에 명시한다. (?i) 같은 인라인 플래그는 JavaScript가
        # 컴파일하지 못해 두 구현이 갈리므로 쓰지 않는다.
        self._patterns = [
            (spec["name"], re.compile(spec["pattern"], _flags(spec.get("flags", ""))), spec.get("reason", ""))
            for spec in data.get("forbidden_patterns") or ()
        ]
        self._field_re: dict[str, re.Pattern] = {}

    @classmethod
    def from_config(cls, config: Config | None = None) -> "EgressGuard":
        config = config or get_config()
        return cls(load_policy(config.policy_dir / "egress-policy.json"))

    # -- 정규식 캐시 -------------------------------------------------------

    def _compiled(self, pattern: str) -> re.Pattern:
        if pattern not in self._field_re:
            self._field_re[pattern] = re.compile(pattern)
        return self._field_re[pattern]

    # -- 1·2단계: 스키마와 값 ------------------------------------------------

    def _check_scalar(self, spec: dict[str, Any], value: Any, path: str) -> list[Violation]:
        out: list[Violation] = []
        expected = spec.get("type", "string")

        if value is None:
            if not spec.get("nullable"):
                out.append(Violation("type", path, "값이 없는데 nullable이 아님"))
            return out

        if expected == "number":
            if isinstance(value, bool) or not isinstance(value, (int, float)):
                out.append(Violation("type", path, f"number가 아님 ({type(value).__name__})"))
                return out
            if "min" in spec and value < spec["min"]:
                out.append(Violation("range", path, f"{value} < 최솟값 {spec['min']}"))
            if "max" in spec and value > spec["max"]:
                out.append(Violation("range", path, f"{value} > 최댓값 {spec['max']}"))
            return out

        if not isinstance(value, str):
            out.append(Violation("type", path, f"string이 아님 ({type(value).__name__})"))
            return out

        if "max_length" in spec and len(value) > spec["max_length"]:
            out.append(Violation("range", path, f"길이 {len(value)} > {spec['max_length']}"))
        if "enum" in spec and value not in spec["enum"]:
            out.append(Violation("pattern", path, f"허용 값이 아님: {value[:40]!r}", value[:40]))
        if "pattern" in spec and not self._compiled(spec["pattern"]).match(value):
            out.append(
                Violation("pattern", path, f"형식 불일치: {spec['pattern']}", value[:60])
            )
        return out

    def _check_field(self, name: str, spec: dict[str, Any], value: Any, path: str) -> list[Violation]:
        expected = spec.get("type", "string")

        if expected == "array":
            if not isinstance(value, list):
                return [Violation("type", path, f"array가 아님 ({type(value).__name__})")]
            out: list[Violation] = []
            if "max_items" in spec and len(value) > spec["max_items"]:
                out.append(Violation("range", path, f"항목 {len(value)} > {spec['max_items']}"))
            item_spec = spec.get("items") or {}
            for index, item in enumerate(value):
                item_path = f"{path}[{index}]"
                if item_spec.get("type") == "object":
                    out.extend(self._check_object(item_spec, item, item_path))
                else:
                    out.extend(self._check_scalar(item_spec, item, item_path))
            return out

        return self._check_scalar(spec, value, path)

    def _check_object(self, spec: dict[str, Any], value: Any, path: str) -> list[Violation]:
        if not isinstance(value, dict):
            return [Violation("type", path, f"object가 아님 ({type(value).__name__})")]
        fields = spec.get("fields") or {}
        out: list[Violation] = []
        for key in value:
            if key.lower() in self.forbidden_names:
                out.append(Violation("forbidden_field", f"{path}.{key}", "금칙 필드명"))
            elif key not in fields:
                out.append(Violation("unknown_field", f"{path}.{key}", "허용 목록에 없는 키"))
        for key, field_spec in fields.items():
            if key in value:
                out.extend(self._check_scalar(field_spec, value[key], f"{path}.{key}"))
        return out

    def _check_schema(self, fact: dict[str, Any], path: str) -> list[Violation]:
        out: list[Violation] = []
        if not isinstance(fact, dict):
            return [Violation("type", path, "VulnFact가 object가 아님")]

        for key in fact:
            if key.lower() in self.forbidden_names:
                # 조립기가 넣을 수 없는 이름이다. 여기 걸렸다면 설계가 어긋난 것이다.
                out.append(Violation("forbidden_field", f"{path}.{key}", "금칙 필드명 — 내부 정보일 수 있음"))
            elif key not in self.allowed:
                out.append(Violation("unknown_field", f"{path}.{key}", "허용 목록에 없는 키"))

        for key, spec in self.allowed.items():
            if key not in fact:
                if spec.get("required"):
                    out.append(Violation("type", f"{path}.{key}", "필수 필드 누락"))
                continue
            out.extend(self._check_field(key, spec, fact[key], f"{path}.{key}"))
        return out

    # -- 3단계: 금칙 패턴 ---------------------------------------------------

    def _walk_strings(self, node: Any, path: str):
        if isinstance(node, str):
            yield path, node
        elif isinstance(node, dict):
            for key, value in node.items():
                yield from self._walk_strings(value, f"{path}.{key}")
        elif isinstance(node, list):
            for index, value in enumerate(node):
                yield from self._walk_strings(value, f"{path}[{index}]")

    def _check_patterns(self, fact: Any, path: str) -> list[Violation]:
        out: list[Violation] = []
        for value_path, text in self._walk_strings(fact, path):
            for name, pattern, reason in self._patterns:
                match = pattern.search(text)
                if match:
                    out.append(
                        Violation("forbidden_pattern", value_path, f"{name}: {reason}", match.group(0)[:60])
                    )
        return out

    # -- 공개 진입점 --------------------------------------------------------

    def check(self, facts: list[dict[str, Any]]) -> SanitizeResult:
        """검증만 하고 결과를 돌려준다. 예외를 던지지 않는다 (미리보기용)."""
        violations: list[Violation] = []

        max_facts = self.limits.get("max_facts_per_request")
        if max_facts and len(facts) > max_facts:
            violations.append(
                Violation("range", "facts", f"한 요청 최대 {max_facts}건인데 {len(facts)}건")
            )

        for index, fact in enumerate(facts):
            path = f"facts[{index}]"
            violations.extend(self._check_schema(fact, path))
            violations.extend(self._check_patterns(fact, path))

        max_bytes = self.limits.get("max_payload_bytes")
        if max_bytes:
            size = len(json.dumps(facts, ensure_ascii=False).encode("utf-8"))
            if size > max_bytes:
                violations.append(Violation("range", "facts", f"크기 {size}B > 한도 {max_bytes}B"))

        return SanitizeResult(
            ok=not violations,
            facts=facts,
            violations=violations,
            policy_version=self.policy.version,
            policy_sha256=self.policy.sha256,
        )

    def enforce(self, facts: list[dict[str, Any]]) -> list[dict[str, Any]]:
        """검증에 실패하면 EgressBlocked를 던진다. 실제 전송 직전에 쓴다.

        '문제가 있는 항목만 빼고 나머지는 보낸다'를 하지 않는 이유: 무엇이
        어떻게 새려 했는지 모르는 상태에서 나머지가 안전하다고 볼 근거가 없다.
        """
        result = self.check(facts)
        if not result.ok:
            raise EgressBlocked(result.violations)
        return result.facts

File: web/js/core/sanitizer.js
/**
 * 이그레스 가드 (브라우저 구현).
 *
 * core/sanitizer.py 와 **같은 정책 파일**(policy/egress-policy.json)을 읽고
 * 같은 판정을 내야 한다. 두 구현의 동치성은 policy/egress-test-vectors.json
 * 으로 검증한다 (tests/js/sanitizer.test.mjs, tests/test_egress.py).
 *
 * 브라우저에도 가드를 두는 이유: GitHub Pages 데모에서 방문자가 자기 키로
 * Gemini를 직접 호출할 때, 서버를 거치지 않으므로 이 구현이 유일한 방어다.
 */

let cachedPolicy = null;

export async function loadEgressPolicy(url = '/policy/egress-policy.json') {
  if (cachedPolicy) return cachedPolicy;
  const response = await fetch(url);
  if (!response.ok) throw new Error(`이그레스 정책을 읽지 못했습니다 (${response.status})`);
  cachedPolicy = await response.json();
  return cachedPolicy;
}

/** JSON 정책의 sha256. Python 쪽 canonical_hash 와 같은 방식(키 정렬·공백 제거). */
export async function policyHash(policy) {
  const stripped = stripComments(policy);
  const canonical = stableStringify(stripped);
  const digest = await crypto.subtle.digest('SHA-256', new TextEncoder().encode(canonical));
  return [...new Uint8Array(digest)].map((b) => b.toString(16).padStart(2, '0')).join('');
}

function stripComments(node) {
  if (Array.isArray(node)) return node.map(stripComments);
  if (node && typeof node === 'object') {
    const out = {};
    for (const [k, v] of Object.entries(node)) {
      if (!k.startsWith('_')) out[k] = stripComments(v);
    }
    return out;
  }
  return node;
}

/** 키를 정렬하고 공백 없이 직렬화한다 (Python json.dumps(sort_keys=True, separators=(',',':'))와 동일). */
function stableStringify(node) {
  if (node === null) return 'null';
  if (Array.isArray(node)) return `[${node.map(stableStringify).join(',')}]`;
  if (typeof node === 'object') {
    const keys = Object.keys(node).sort();
    return `{${keys.map((k) => `${JSON.stringify(k)}:${stableStringify(node[k])}`).join(',')}}`;
  }
  return JSON.stringify(node);
}

export class EgressGuard {
  constructor(policy) {
    this.policy = policy;
    this.allowed = policy.allowed_fields || {};
    this.forbiddenNames = new Set((policy.forbidden_field_names || []).map((n) => n.toLowerCase()));
    this.limits = policy.limits || {};
    // 플래그는 정책에 명시된 것만 쓴다. (?i) 같은 인라인 플래그는 JS가
    // 컴파일하지 못하므로 정책에서 금지되어 있다.
    this.patterns = (policy.forbidden_patterns || []).map((spec) => ({
      name: spec.name,
      re: new RegExp(spec.pattern, spec.flags || ''),
      reason: spec.reason || '',
    }));
    this._fieldRe = new Map();
  }

  static async load(url) {
    return new EgressGuard(await loadEgressPolicy(url));
  }

  _re(pattern) {
    if (!this._fieldRe.has(pattern)) this._fieldRe.set(pattern, new RegExp(pattern));
    return this._fieldRe.get(pattern);
  }

  // --- 1·2단계: 스키마와 값 -------------------------------------------------

  _checkScalar(spec, value, path) {
    const out = [];
    const expected = spec.type || 'string';

    if (value === null || value === undefined) {
      if (!spec.nullable) out.push({ rule: 'type', path, reason: '값이 없는데 nullable이 아님', sample: '' });
      return out;
    }

    if (expected === 'number') {
      if (typeof value !== 'number' || Number.isNaN(value)) {
        out.push({ rule: 'type', path, reason: `number가 아님 (${typeof value})`, sample: '' });
        return out;
      }
      if (spec.min !== undefined && value < spec.min)
        out.push({ rule: 'range', path, reason: `${value} < 최솟값 ${spec.min}`, sample: '' });
      if (spec.max !== undefined && value > spec.max)
        out.push({ rule: 'range', path, reason: `${value} > 최댓값 ${spec.max}`, sample: '' });
      return out;
    }

    if (typeof value !== 'string') {
      out.push({ rule: 'type', path, reason: `string이 아님 (${typeof value})`, sample: '' });
      return out;
    }

    if (spec.max_length !== undefined && value.length > spec.max_length)
      out.push({ rule: 'range', path, reason: `길이 ${value.length} > ${spec.max_length}`, sample: '' });
    if (spec.enum && !spec.enum.includes(value))
      out.push({ rule: 'pattern', path, reason: `허용 값이 아님: ${value.slice(0, 40)}`, sample: value.slice(0, 40) });
    if (spec.pattern) {
      const m = this._re(spec.pattern).exec(value);
      // Python의 re.match 는 문자열 시작에서만 맞춘다. JS도 같게 맞춘다.
      if (!m || m.index !== 0) {
        out.push({ rule: 'pattern', path, reason: `형식 불일치: ${spec.pattern}`, sample: value.slice(0, 60) });
      }
    }
    return out;
  }

  _checkObject(spec, value, path) {
    if (value === null || typeof value !== 'object' || Array.isArray(value)) {
      return [{ rule: 'type', path, reason: `object가 아님 (${typeof value})`, sample: '' }];
    }
    const fields = spec.fields || {};
    const out = [];
    for (const key of Object.keys(value)) {
      if (this.forbiddenNames.has(key.toLowerCase()))
        out.push({ rule: 'forbidden_field', path: `${path}.${key}`, reason: '금칙 필드명', sample: '' });
      else if (!(key in fields))
        out.push({ rule: 'unknown_field', path: `${path}.${key}`, reason: '허용 목록에 없는 키', sample: '' });
    }
    for (const [key, fieldSpec] of Object.entries(fields)) {
      if (key in value) out.push(...this._checkScalar(fieldSpec, value[key], `${path}.${key}`));
    }
    return out;
  }

  _checkField(spec, value, path) {
    if ((spec.type || 'string') === 'array') {
      if (!Array.isArray(value))
        return [{ rule: 'type', path, reason: `array가 아님 (${typeof value})`, sample: '' }];
      const out = [];
      if (spec.max_items !== undefined && value.length > spec.max_items)
        out.push({ rule: 'range', path, reason: `항목 ${value.length} > ${spec.max_items}`, sample: '' });
      const itemSpec = spec.items || {};
      value.forEach((item, index) => {
        const itemPath = `${path}[${index}]`;
        out.push(...(itemSpec.type === 'object'
          ? this._checkObject(itemSpec, item, itemPath)
          : this._checkScalar(itemSpec, item, itemPath)));
      });
      return out;
    }
    return this._checkScalar(spec, value, path);
  }

  _checkSchema(fact, path) {
    const out = [];
    if (fact === null || typeof fact !== 'object' || Array.isArray(fact)) {
      return [{ rule: 'type', path, reason: 'VulnFact가 object가 아님', sample: '' }];
    }
    for (const key of Object.keys(fact)) {
      if (this.forbiddenNames.has(key.toLowerCase()))
        out.push({ rule: 'forbidden_field', path: `${path}.${key}`, reason: '금칙 필드명 — 내부 정보일 수 있음', sample: '' });
      else if (!(key in this.allowed))
        out.push({ rule: 'unknown_field', path: `${path}.${key}`, reason: '허용 목록에 없는 키', sample: '' });
    }
    for (const [key, spec] of Object.entries(this.allowed)) {
      if (!(key in fact)) {
        if (spec.required) out.push({ rule: 'type', path: `${path}.${key}`, reason: '필수 필드 누락', sample: '' });
        continue;
      }
      out.push(...this._checkField(spec, fact[key], `${path}.${key}`));
    }
    return out;
  }

  // --- 3단계: 금칙 패턴 -----------------------------------------------------

  *_walkStrings(node, path) {
    if (typeof node === 'string') yield [path, node];
    else if (Array.isArray(node)) {
      for (let i = 0; i < node.length; i += 1) yield* this._walkStrings(node[i], `${path}[${i}]`);
    } else if (node && typeof node === 'object') {
      for (const [k, v] of Object.entries(node)) yield* this._walkStrings(v, `${path}.${k}`);
    }
  }

  _checkPatterns(fact, path) {
    const out = [];
    for (const [valuePath, text] of this._walkStrings(fact, path)) {
      for (const { name, re, reason } of this.patterns) {
        const m = re.exec(text);
        if (m) {
          out.push({
            rule: 'forbidden_pattern', path: valuePath,
            reason: `${name}: ${reason}`, sample: m[0].slice(0, 60),
          });
        }
      }
    }
    return out;
  }

  // --- 공개 진입점 ----------------------------------------------------------

  check(facts) {
    const violations = [];
    const maxFacts = this.limits.max_facts_per_request;
    if (maxFacts && facts.length > maxFacts) {
      violations.push({ rule: 'range', path: 'facts',
        reason: `한 요청 최대 ${maxFacts}건인데 ${facts.length}건`, sample: '' });
    }

    facts.forEach((fact, index) => {
      const path = `facts[${index}]`;
      violations.push(...this._checkSchema(fact, path));
      violations.push(...this._checkPatterns(fact, path));
    });

    const maxBytes = this.limits.max_payload_bytes;
    if (maxBytes) {
      const size = new TextEncoder().encode(JSON.stringify(facts)).length;
      if (size > maxBytes)
        violations.push({ rule: 'range', path: 'facts', reason: `크기 ${size}B > 한도 ${maxBytes}B`, sample: '' });
    }

    return { ok: violations.length === 0, facts, violations };
  }

  /** 검증에 실패하면 던진다. 일부만 골라 보내지 않는다 — 전부 아니면 아무것도. */
  enforce(facts) {
    const result = this.check(facts);
    if (!result.ok) {
      const head = result.violations.slice(0, 5)
        .map((v) => `${v.rule}@${v.path}: ${v.reason}`).join('; ');
      const error = new Error(`외부 전송이 차단되었습니다 (${head})`);
      error.violations = result.violations;
      throw error;
    }
    return result.facts;
  }
}
